Sources/AsyncAlgorithms/AsyncRecursiveMapSequence.swift

```diff
@@ -0,0 +1,127 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift Async Algorithms open source project
+//
+// Copyright (c) 2022 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+//
+//===----------------------------------------------------------------------===//
+
+extension AsyncSequence {
+    /// Returns a sequence containing the original sequence followed by recursive mapped sequence.
+    ///
+    /// ```
+    /// struct View {
+    ///     var id: Int
+    ///     var children: [View] = []
+    /// }
+    /// let tree = [
+    ///     View(id: 1, children: [
+    ///         View(id: 3),
+    ///         View(id: 4, children: [
+    ///             View(id: 6),
+    ///         ]),
+    ///         View(id: 5),
+    ///     ]),
+    ///     View(id: 2),
+    /// ]
+    /// for await view in tree.async.recursiveMap({ $0.children.async }) {
+    ///     print(view.id)
+    /// }
+    /// // 1
+    /// // 2
+    /// // 3
+    /// // 4
+    /// // 5
+    /// // 6
+    /// ```
+    ///
+    /// - Parameters:
+    ///   - transform: A closure that map the element to new sequence.
+    /// - Returns: A sequence of the original sequence followed by recursive mapped sequence.
+    @inlinable
+    public func recursiveMap<C>(_ transform: @Sendable @escaping (Element) async -> C) -> AsyncRecursiveMapSequence<Self, C> {
+        return AsyncRecursiveMapSequence(self, transform)
+    }
+}
+
+public struct AsyncRecursiveMapSequence<Base: AsyncSequence, Transformed: AsyncSequence>: AsyncSequence where Base.Element == Transformed.Element {
+    
+    public typealias Element = Base.Element
+    
+    @usableFromInline
+    let base: Base
+    
+    @usableFromInline
+    let transform: @Sendable (Base.Element) async -> Transformed
+    
+    @inlinable
+    init(_ base: Base, _ transform: @Sendable @escaping (Base.Element) async -> Transformed) {
+        self.base = base
+        self.transform = transform
+    }
+    
+    @inlinable
+    public func makeAsyncIterator() -> AsyncIterator {
+        return AsyncIterator(base, transform)
+    }
+}
+
+extension AsyncRecursiveMapSequence {
+    
+    public struct AsyncIterator: AsyncIteratorProtocol {
+        
+        @usableFromInline
+        var base: Base.AsyncIterator?
+        
+        @usableFromInline
+        var mapped: ArraySlice<Transformed> = []
```

Comment:
Yes, it’s definitely problems. Maybe I should change to depth first algorithm and it shouldn’t have indefinite depth of tree hopefully.

Reply:
This is pretty cool!

I have a somewhat and open-ended question, I believe using depth first would change the expected output in the example above from 1,2,3,4,5,6 to 1,3,4,6,5,2. The traversal would still be correct but given that the output varies so widely depending on the type of traversal, should that be communicated in the function name? Or is order not that important in this scenario? 

Is the utility of this function its ability to visit each node once and iterate over the tree or do/should we care about the order in the sequence?

Reply:
This function is a simplified version of SQL recursive CTE, I implemented breadth-first traversal because the form of the algorithm is already in my head.
It can be done by function name, or using a enum option for choosing the traversal methods.